moveit_ros/moveit_servo/src/servo.cpp

```diff
@@ -33,74 +33,421 @@
+  // Create Eigen maps for cleaner operations.
+  Eigen::Map<Eigen::VectorXd> current_joint_positions(current_state.positions.data(), num_joints_);
+  Eigen::Map<Eigen::VectorXd> target_joint_positions(target_state.positions.data(), num_joints_);
+  Eigen::Map<Eigen::VectorXd> current_joint_velocities(current_state.velocities.data(), num_joints_);
+  Eigen::Map<Eigen::VectorXd> target_joint_velocities(target_state.velocities.data(), num_joints_);
+
+  // Update filter state
+  smoother_->reset(current_state.positions);
+
+  // Compute the change in joint position due to the incoming command
+  Eigen::VectorXd joint_position_delta = jointDeltaFromCommand(command);
+
+  // Continue rest of the computations only if the command was valid
+  if (servo_status_ != StatusCode::INVALID)
+  {
+    // Apply collision scaling to the joint position delta
+    if (collision_velocity_scale_ > 0 && collision_velocity_scale_ < 1)
+    {
+      servo_status_ = StatusCode::DECELERATE_FOR_COLLISION;
+    }
+    else if (collision_velocity_scale_ == 0)
+    {
+      servo_status_ = StatusCode::HALT_FOR_COLLISION;
+    }
+    joint_position_delta *= collision_velocity_scale_;
+
+    // Compute the next joint positions based on the joint position deltas
+    target_joint_positions = current_joint_positions + joint_position_delta;
+
+    // TODO : apply filtering to the velocity instead of position
+    // Apply smoothing to the positions
+    smoother_->doSmoothing(target_state.positions);
+
+    // Compute velocities based on smoothed joint positions
+    target_joint_velocities = (target_joint_positions - current_joint_positions) / servo_params_.publish_period;
+
+    // Scale down the velocity based on joint velocity limit or user defined scaling if applicable.
+    target_joint_velocities *=
+        velocityScalingFactor(target_joint_velocities, joint_bounds_, servo_params_.override_velocity_scaling_factor);
+
+    // Adjust joint position based on scaled down velocity
+    target_joint_positions = current_joint_positions + (target_joint_velocities * servo_params_.publish_period);
+
+    // Check if any joints are going past joint position limits
+    std::vector<int> joints_to_halt =
+        jointsToHalt(target_joint_positions, target_joint_velocities, joint_bounds_, servo_params_.joint_limit_margin);
+
+    // Apply halting if any joints need to be halted.
+    if (!joints_to_halt.empty())
+    {
+      servo_status_ = StatusCode::JOINT_BOUND;
+      target_state = haltJoints(joints_to_halt, current_state, target_state);
+    }
+  }
+
+  return target_state;
+}
+
+Eigen::VectorXd Servo::jointDeltaFromCommand(const ServoInput& command)
+{
+  Eigen::VectorXd target_joint_positions(num_joints_);
+  robot_state_->copyJointGroupPositions(joint_model_group_, target_joint_positions);
+
+  const CommandType incoming_type = incomingCommandType();
+
+  if (incoming_type == CommandType::JOINT_JOG && command.index() == static_cast<size_t>(incoming_type))
```

Comment:
Since the `command.index()` check is common to all of the cases, you could pull this out one level above, so

```
if (command.index() == static_cast<size_t>(incoming_type))
{
  if (incoming_type == CommandType::JOINT_JOG)
  {
    ...
```

Reply:
Good catch!
This helped make the error message more specific as well.